cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -1986,3 +1986,76 @@
 
 (rule (lower (has_type (fits_in_64 ty) (breduce src)))
       (value_regs_get_gpr src 0))
+
+;; Rules for `bint` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; Booleans are stored as all-zeroes (0) or all-ones (-1). We AND out
+;; the LSB to give a 0 / 1-valued integer result.
+
+(rule (lower (has_type (fits_in_64 ty)
+                       (bint src)))
+      (x64_and ty src (RegMemImm.Imm 1)))
+(rule (lower (has_type $I128
+                       (bint src)))
+      (value_regs
+       (x64_and $I64 src (RegMemImm.Imm 1))
+       (imm $I64 0)))
+
+;; Rules for `debugtrap` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(rule (lower (debugtrap))
+      (side_effect (hlt)))
```

Comment:
Existing/curious: Do you know why `debugtrap` isn't a safe point? Our other traps are safe points.

Reply:
I'm actually not sure! For what it's worth, `hlt` is a bit of an odd choice here as well (it can only execute in kernel mode, and I think generates `SIGILL` in userspace). I guess the idea is that it's not meant to be a "normal" transfer of control to the runtime paired with the generated code, but rather a place where a debugger could catch a signal or hook in with `ptrace` or something of the sort, so setting up state for a "normal" runtime invocation to e.g. scan the stack wouldn't be useful. I'm not sure though.

In the absence of more context here I'm inclined to leave the lowering as-is (translated from existing), but we should definitely ask whether this opcode is even necessary, or at least what its precise guarantees should be, when we get to cleaning up the CLIF opcode space...